peps/pep-0817.rst

```diff
@@ -0,0 +1,2528 @@
+a given wheel's supported environments. These tags are unable to express
+features of modern hardware configurations, such as the availability of
+GPU acceleration, or to provide custom package variants, such as builds
+against different dependency ABIs. This is particularly challenging for
+the scientific computing, artificial intelligence (AI), machine learning
+(ML), and high-performance computing (HPC) communities.
+
+This PEP proposes "Wheel Variants", an extension to the
+:doc:`packaging:specifications/binary-distribution-format`. This
+extension introduces a mechanism for package maintainers to declare
+multiple build variants for the same package version, while allowing
+installers to automatically select the most appropriate variant based on
+system hardware and software characteristics. More specifically, it
+proposes:
+
+- An evolution of the wheel format called **Wheel Variant** that allows
+  wheels to be distinguished by hardware or software attributes.
+
+- A **variant provider plugin** interface that allows installers to
+  dynamically detect platform attributes and select the most suitable
+  wheel.
+
+The goal is for a familiar ``[uv] pip install <package>`` to provide
+the best user experience.
+
+
+Motivation
+==========
+
+The Python packaging ecosystem has evolved to support increasingly
+diverse computing environments. The current software ecosystem often
+relies on platform-specific features to determine which binaries are
+compatible with a particular machine. Unfortunately the current wheel
+format cannot adequately express the diversity of features present
+in modern hardware.
+
+For example, packages such as `PyTorch <https://pytorch.org/>`_ need to
+be built for specific CUDA or ROCm versions, and that information cannot
+currently be included in the wheel tag. Having to build multiple wheels
+targeting very different hardware configurations forces maintainers into
+various distribution strategies that are suboptimal, and create friction
+for users and authors of other software who wish to depend on the
+package in question.
+
+A few existing approaches are explored in the subsequent subsections.
+These include maintaining separate package indexes for different
+hardware configurations, bundling all potential variants into a single
+wheel of considerable size, or using separate package names
+(``mypackage-gpu``, ``mypackage-cpu``, etc.). Each of these approaches
+has significant drawbacks.
+
+According to the `2024 Python Developers Survey
+<https://lp.jetbrains.com/python-developers-survey-2024/#purposes-for-using-python>`__,
+a significant portion of respondents over the last years have been
+successively using Python for scientific computing purposes, covering
+such areas as Data analysis (steadily over 40% respondents), Machine
+learning (grown to 40% in 2024) and Data engineering (around 30%).
+Many of these use cases are directly impacted by suboptimal
+packaging.
```

Comment:
I'd make this para the start of Motivation, it provides a useful contextualisation
```suggestion
The `2024 Python Developers Survey`__ shows that a significant proportion of
Python's users have scientific computing use-cases. This includes data analysis
(40% of respondents), machine learning (30%), and data engineering (30%).
Many of these use-cases are directly impacted by suboptimal Python packaging.

__ https://lp.jetbrains.com/python-developers-survey-2024/#purposes-for-using-python
```